index.bs

```diff
@@ -4648,71 +4659,151 @@ The <dfn export for=commands>browsingContext.setViewport</dfn> command modifies
+<div algorithm>
+To <dfn>set device pixel ratio override</dfn> given |navigable| and |device pixel ratio|:
 
-The [=remote end steps=] with |command parameters| are:
+1. If |device pixel ratio| is not null:
 
-1. Let |navigable id| be the value of the <code>context</code> field of |command
-   parameters|.
+   1. When the [=select an image source from a source set=] steps are run, act as if
+      the implementation's pixel density was set to |device pixel ratio| when selecting an image.
 
-1. Let |navigable| be the result of [=trying=] to [=get a navigable=] with
-   |navigable id|.
+   1. For the purposes of the [=resolution media feature=], act as if
+      the implementation's resolution is |device pixel ratio| dppx scaled by the page zoom.
 
-1. If |navigable| is not a [=/top-level traversable=], return [=error=] with
-   [=error code=] [=invalid argument=].
+   1. [=map/Set=] [=device pixel ratio overrides=][|navigable|] to |device pixel ratio|.
+
+      Note: This will take an effect because of the patch of [[#patchs-determine-the-device-pixel-ratio]].
+
+1. Otherwise:
+
+   1. When the [=select an image source from a source set=] steps are run, use the implementation's default behavior,
+      without any changes made by previous invocations of these steps.
+
+   1. For the purposes of the [=resolution media feature=], use the implementation's default behavior,
+      without any changes made by previous invocations of these steps.
+
+   1. [=map/Remove=] |navigable| from [=device pixel ratio overrides=].
+
+1. Run [=evaluate media queries and report changes=] for [=/document=] currently loaded
+   in a specified |navigable|.
+
+</div>
+
+
+<div algorithm>
+To <dfn>set viewport</dfn> given |navigable| and |viewport|:
+
+1. If |viewport| is not null, set the width of |navigable|'s [=layout
+   viewport=] to be the <code>width</code> of |viewport| in CSS pixels and
+   set the height of the |navigable|'s [=layout viewport=] to be the
+   <code>height</code> of |viewport| in CSS pixels.
+
+1. Otherwise, set the |navigable|'s [=layout viewport=] to the
+   implementation-defined default.
+
+</div>
+
+<div algorithm="set viewport when a top-level traversable is created">
+
+After creating a document in a new [=navigable/top-level traversable=] |navigable| and
+before the [=run WebDriver BiDi preload scripts=] algorithm is invoked:
+
+1. Let |user context| be |navigable|'s [=associated user context=].
+
+1. If [=viewport overrides map=] [=map/contains=] |user context|:
+
+   1. If the <code>viewport</code> field of [=viewport overrides map=][|user context|] is not null:
 
-1. If the implementation is unable to adjust the |navigable|'s [=layout viewport=]
+      1. [=Set viewport=] with |navigable| and [=viewport overrides map=][|user context|]["<code>viewport</code>"].
+
+   1. If the <code>devicePixelRatio</code> field of [=viewport overrides map=][|user context|] is not null:
+
+      1. For the |navigable| and all [=descendant navigables=]:
```

Comment:
I guess the new descendant navigables created after the top level navigable is created will not have the "devicePixelRatio" updated.

Reply:
Yeah, James also pointed it out, I've updated the algorithm to apply to all kind of navigables, so it should hopefully cover it